backend/src/apiserver/storage/pipeline_store_kubernetes.go

```diff
@@ -0,0 +1,519 @@
+
+	err := k.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, &k8sPipeline)
+	if k8serrors.IsNotFound(err) {
+		return nil, util.NewResourceNotFoundError("Namespace/Pipeline", fmt.Sprintf("%v/%v", namespace, name))
+	} else if err != nil {
+		return nil, util.NewInternalServerError(
+			err, "Failed to get a pipeline with name %v and namespace %v", name, namespace,
+		)
+	}
+
+	return k8sPipeline.ToModel(), nil
+}
+
+func (k *PipelineStoreKubernetes) ListPipelinesV1(filterContext *model.FilterContext, opts *list.Options) ([]*model.Pipeline, []*model.PipelineVersion, int, string, error) {
+	return nil, nil, 0, "", ErrNoV1
+}
+
+func (k *PipelineStoreKubernetes) ListPipelines(filterContext *model.FilterContext, opts *list.Options) ([]*model.Pipeline, int, string, error) {
+	k8sPipelines := v2beta1.PipelineList{}
+
+	listOptions := []ctrlclient.ListOption{ctrlclient.UnsafeDisableDeepCopy}
+
+	if filterContext.ReferenceKey != nil && filterContext.ReferenceKey.Type == model.NamespaceResourceType {
+		listOptions = append(listOptions, ctrlclient.InNamespace(filterContext.ReferenceKey.ID))
+	}
+
+	// Be careful, the deep copy is disabled here to reduce memory allocations
+	err := k.client.List(context.TODO(), &k8sPipelines, listOptions...)
+	if err != nil {
+		return nil, 0, "", util.NewInternalServerError(
+			err, "Failed to find the pipeline associated with this pipeline version",
+		)
+	}
+
+	pipelines := make([]*model.Pipeline, 0, len(k8sPipelines.Items))
+
+	for _, k8sPipeline := range k8sPipelines.Items {
+		if opts.Filter == nil {
+			pipelines = append(pipelines, k8sPipeline.ToModel())
+			continue
+		} else {
+			found, err1 := opts.Filter.FilterK8sPipelines(k8sPipeline)
+			if err1 != nil {
+				return nil, 0, "", err1
+			}
+			if found {
+				pipelines = append(pipelines, k8sPipeline.ToModel())
+			}
+		}
+	}
+
+	// Because controller-client does not have sorting, use this function to sort by fields.
+	sort.Slice(pipelines, func(i, j int) bool {
+		if strings.Compare(opts.SortByFieldName, "CreatedAtInSec") == 0 {
+			return pipelines[i].CreatedAtInSec > pipelines[j].CreatedAtInSec
+		} else if strings.Compare(opts.SortByFieldName, "Name") == 0 {
+			return pipelines[i].Name > pipelines[j].Name
+		}
+		return false
+	})
+	if !opts.IsDesc {
+		slices.Reverse(pipelines)
+	}
+
+	// If there's pagination, find the index of the first element of the next page
+	nextPage := 0
+	for p, pipeline := range pipelines {
+		if strings.Compare(fmt.Sprint(opts.KeyFieldValue), pipeline.UUID) == 0 {
+			nextPage = p
+
+			break
+		}
+	}
+
+	// Split results on multiple pages if needed
+	upperBound := nextPage + opts.PageSize
+	npt := ""
+	if len(pipelines) <= opts.PageSize {
+		return pipelines, len(pipelines), "", nil
+	} else if upperBound >= len(pipelines) {
+		upperBound = len(pipelines)
+		npt, err = opts.NextPageToken(pipelines[upperBound-1])
+	} else {
+		npt, err = opts.NextPageToken(pipelines[upperBound])
+	}
```

Comment:
Would this simplification work?
```suggestion
	if len(pipelines) >= upperBound {
		npt, err = opts.NextPageToken(pipelines[upperBound])
	}
```